### ISSUE TITLE
Bean Class in Java

### ISSUE THREAD

**Issue by anitsh** [codeanit/todo#971]
JavaBeans are classes that encapsulate many objects into a single object (the bean). It is a java class that should follow following conventions: Must implement Serializable. It should have a public no-arg constructor. All properties in java bean must be private with public getters and setter methods.

The name "Bean" was given to encompass this standard, which aims to create reusable software components for Java. Spring Bean: is an object, which is created, managed and destroyed in Spring Container.

Bean is a key concept of the Spring Framework. So understanding this notion is crucial to get the hang of the framework and use it in an effective way.

Unfortunately, there aren't clear answers to the simple question of what a Spring bean really is. Some explanations go to such a low level that the big picture is missed, whereas others are too vague.

A definition of beans in [the Spring Framework documentation](https://docs.spring.io/spring/docs/current/spring-framework-reference/core.html#beans-introduction):

In Spring, the objects that form the backbone of your application and that are managed by the Spring IoC container are called beans.

A bean is an object that is instantiated, assembled, and otherwise managed by a Spring IoC container.

Inversion of Control (IoC) is a process in which an object defines its dependencies without creating them. This object delegates the job of constructing such dependencies to an IoC container.

Instead of constructing dependencies by itself, an object can retrieve its dependencies from an IoC container. All we need to do is to provide the container with appropriate configuration metadata.

## Traditional Approach
public class Company {
    private Address address;

    public Company(Address address) {
        this.address = address;
    }

    // getter, setter and other properties
}

public class Address {
    private String street;
    private int number;

    public Address(String street, int number) {
        this.street = street;
        this.number = number;
    }

    // getters and setters
}

Address address = new Address("High Street", 1000);
Company company = new Company(address);

## Spring IoC Approach

Decorate the Company class with the @Component annotation.
@Component
public class Company {
    // this body is the same as before
}

A configuration class supplying bean metadata to an IoC container. The configuration class produces a bean of type Address. 

It also carries the @ComponentScan annotation, which instructs the container to look for beans in the package containing the Company class.

When a Spring IoC container constructs objects of those types, all the objects are called Spring beans, as they are managed by the IoC container.

@Configuration
@ComponentScan(basePackageClasses = Company.class)
public class Config {
    @Bean
    public Address getAddress() {
        return new Address("High Street", 1000);
    }
}

Since we defined beans in a configuration class, we'll need an instance of the `AnnotationConfigApplicationContext` class to build up a container:
`ApplicationContext context = new AnnotationConfigApplicationContext(Config.class);`

A JavaBean is just a [standard](http://www.oracle.com/technetwork/java/javase/documentation/spec-136004.html). It is a regular Java class, except it follows certain conventions:

All properties are private (use [getters/setters](http://en.wikipedia.org/wiki/Mutator_method))
A public [no-argument constructor](http://en.wikipedia.org/wiki/Nullary_constructor)
Implements [Serializable](http://docs.oracle.com/javase/8/docs/api/java/io/Serializable.html).
That's it. It's just a convention. Lots of libraries depend on it though.

With respect to Serializable, from the [API documentation](http://docs.oracle.com/javase/8/docs/api/java/io/Serializable.html):

Serializability of a class is enabled by the class implementing the java.io.Serializable interface. Classes that do not implement this interface will not have any of their state serialized or deserialized. All subtypes of a serializable class are themselves serializable. The serialization interface has no methods or fields and serves only to identify the semantics of being serializable.

In other words, serializable objects can be written to streams, and hence files, object databases, anything really.

Also, there is no syntactic difference between a JavaBean and another class -- a class is a JavaBean if it follows the standards.

There is a term for it, because the standard allows libraries to programmatically do things with class instances you define in a predefined way. For example, if a library wants to stream any object you pass into it, it knows it can because your object is serializable (assuming the library requires your objects be proper JavaBeans).

# Resource
- https://www.baeldung.com/spring-bean
- https://en.wikipedia.org/wiki/JavaBeans
- https://stackoverflow.com/questions/3295496/what-is-a-javabean-exactly
